Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -680,56 +680,86 @@ static bool needsEAGLOnMac()
+    if (mtlDevice != [handle device]) {
+        LOG(WebGL, "MTLSharedTextureHandle does not have the same Metal device as platformDisplay in attachMTLTextureToExternalImage.");
         return std::nullopt;
     }
 
-    auto texture = adoptNS([handle.get().device newSharedTextureWithHandle:handle.get()]);
-    if (!texture) {
-        LOG(WebGL, "Unable to create a MTLSharedTexture from the texture handle in attachIOSurfaceToTexture.");
+    // Create a MTLTexture out of the MTLSharedTextureHandle.
+    auto texture = adoptNS([mtlDevice newSharedTextureWithHandle:handle]);
+    if (!texture)
         return std::nullopt;
-    }
 
     GCGLuint textureWidth = [texture width];
     GCGLuint textureHeight = [texture height];
 
     // FIXME: Does the texture have the correct usage mode?
 
     // Create an EGLImage out of the MTLTexture
-    auto display = platformDisplay();
-    auto eglImage = EGL_CreateImageKHR(display, EGL_NO_CONTEXT, EGL_METAL_TEXTURE_ANGLE, reinterpret_cast<EGLClientBuffer>(texture.get()), emptyAttributes);
-    if (!eglImage) {
-        LOG(WebGL, "Unable to create an EGLImage from the Metal handle in attachIOSurfaceToTexture.");
+    constexpr EGLint emptyAttributes[] = { EGL_NONE };
+    auto eglImage = EGL_CreateImageKHR(platformDisplay, EGL_NO_CONTEXT, EGL_METAL_TEXTURE_ANGLE, reinterpret_cast<EGLClientBuffer>(texture.get()), emptyAttributes);
+    if (!eglImage)
         return std::nullopt;
-    }
 
     // Tell the currently bound texture to use the EGLImage.
     GL_EGLImageTargetTexture2DOES(target, eglImage);
 
-    return std::make_tuple(eglImage, textureWidth, textureHeight);
+    return std::make_tuple(eglImage, IntSize(textureWidth, textureHeight));
 }
 
-void GraphicsContextGLCocoa::detachIOSurfaceFromSharedTexture(void* handle)
+std::optional<ExternalImageAttachResult> GraphicsContextGLCocoa::createAndBindExternalImage(GCGLenum target, IOSurface* surface)
 {
-    auto display = platformDisplay();
-    EGL_DestroyImageKHR(display, handle);
+    MTLSharedTextureHandle* handle = [[MTLSharedTextureHandle alloc] initWithIOSurface:surface->surface() label:@"WebXR"];
+    if (!handle)
+        return std::nullopt;
+
+    return createAndBindSharedTextureToExternalImage(platformDisplay(), target, handle);
 }
 #endif
 
+std::optional<ExternalImageAttachResult> GraphicsContextGLCocoa::createAndBindExternalImage(GCGLenum target, ExternalImageSource source)
+{
+#if PLATFORM(IOS_FAMILY_SIMULATOR)
+    // MTLSharedTexture is unsupported on simulator
+    UNUSED_VARIABLE(target);
+    UNUSED_VARIABLE(source);
+
+    return std::nullopt;
+#else // !PLATFORM(IOS_FAMILY_SIMULATOR
+
+    MTLSharedTextureHandle* handle = nil;
+    return WTF::switchOn(source,
+        [&](const ExternalImageSourceIOSurfaceHandle& ioSurface) -> std::optional<ExternalImageAttachResult> {
+            auto surface = IOSurface::createFromSendRight(WTFMove(ioSurface.handle));
```

Comment:
this looks somehow odd, e.g. you move internals of the const ref object into the IOSurface::createFromSendRight.
This works because of bug in  IOSurface::createFromSendRight type signature.

So you should have form of:
```
return WTF::switchOn(WTFMove(source),
         [&](ExternalImageSourceIOSurfaceHandle&& ioSurface) -> std::optional<ExternalImageAttachResult> {
          ....
        }, 
                 [&](ExternalImageSourceMTLSharedTextureHandle&& sharedTexture) -> std::optional<ExternalImageAttachResult> {
                  ....
       }
```